File: src/Node.py
from dataclasses import dataclass, field
from itertools import chain, product
from typing import Iterable, List


class Node:

    def values(self) -> Iterable[str]:
        raise NotImplemented


@dataclass
class BranchingNode (Node):
    alternatives: List[Node] = field(default_factory=list)

    def values(self) -> Iterable[str]:
        return chain.from_iterable(n.values() for n in self.alternatives)


@dataclass
class SequenceNode (Node):
    sub_rules: List[Node] = field(default_factory=list)

    def values(self) -> Iterable[str]:
        all_values = [n.values() for n in self.sub_rules]
        for values_tuple in product(*all_values):
            yield ' '.join(s for s in values_tuple if s)


@dataclass
class ValueNode (Node):
    value: str

    def values(self) -> Iterable[str]:
        return [self.value]

File: src/Node.js
import { flatten, product } from "./Itertools.js";

class BranchingNode {

    constructor(alternatives = []) {
        this.alternatives = alternatives;
    }

    values() {
        return flatten(this.alternatives.map((alt) => alt.values()));
    }
}

class SequenceNode {

    constructor(sub_rules = []) {
        this.sub_rules = sub_rules;
    }

    values() {
        let result_list = [];
        let cartesian_product = product(this.sub_rules.map((sr) => sr.values()));
        for (let combination of cartesian_product) {
            let res = [...combination].filter((s) => s !== '').join(' ');
            result_list.push(res);
        }
        return result_list;
    }
}

class ValueNode {

    constructor(value) {
        this.value = value;
    }

    values() {
        return [this.value];
    }
}

export { BranchingNode, SequenceNode, ValueNode };